pkg/analyzer/nodes/node.go

```diff
@@ -103,22 +104,57 @@ func extractFilesFromDirectory(root string, matcher matcher.PrefixMatcher) (*fil
 	m := metrics.FileExtractionMetrics{}
 	// TODO(ROX-13771): Use `range matcher.GetCommonPrefixDirs()` again after fixing.
 	for _, dir := range []string{"etc/", "usr/share/rpm", "var/lib/rpm", "usr/share/buildinfo"} {
-		if err := n.addFiles(filepath.FromSlash(dir), matcher, &m); err != nil {
-			return nil, errors.Wrapf(err, "failed to match filesMap at %q (at %q)", dir, n.root)
+		if err := n.addFiles(ctx, filepath.FromSlash(dir), matcher, &m); err != nil {
+			if errors.Is(err, context.DeadlineExceeded) {
+				err = errors.Wrapf(err, "operation took longer than expected and was interrupted")
+			}
+			return nil, errors.Wrapf(err, "failed to extract files from %q at root %q", dir, n.root)
 		}
 	}
 	m.Emit()
 	return n, nil
 }
 
+func walkDirWithContext(ctx context.Context, dir string, fn fs.WalkDirFunc) error {
+	errC := make(chan error)
+	go func(ctx context.Context) {
+		defer close(errC)
+		// this select prevents goroutine leak when ctx times out first
+		select {
+		case <-ctx.Done():
+			return
+		case errC <- filepath.WalkDir(dir, func(path string, info fs.DirEntry, err error) error {
+			if err = fn(path, info, err); err != nil {
+				return err
+			}
+			// this select prevents further calls to fn if ctx was canceled
+			select {
+			case <-ctx.Done():
+				return ctx.Err()
+			default:
+				return nil
+			}
+		}):
+		}
+	}(ctx)
+
+	// Wait for the WalkDir to complete or for the context to be cancelled.
+	select {
+	case <-ctx.Done():
+		return ctx.Err()
+	case err := <-errC:
+		return err
+	}
+}
```

Comment:
Oh, good catch on the goroutine leak.

But now, all this code has become a little bit convoluted. Do you think we can work it out by buffering the channel?

I also removed passing the context to the anonymous goroutine and using the closure because I think it's easier to read it like that, but let me know if it won't work for any reason.

```suggestion
func walkDirWithContext(ctx context.Context, dir string, fn fs.WalkDirFunc) error {
	// Buffering so we can give up on reading the channel and not block the
	// child goroutine.
	errC := make(chan error, 1)

	go func() {
		defer close(errC)
		errC <- filepath.WalkDir(dir, func(path string, info fs.DirEntry, err error) error {
			if err = fn(path, info, err); err != nil {
				return err
			}
			select {
			case <-ctx.Done():
				return ctx.Err()
			default:
				return nil
			}
		})
	}()

	select {
	case <-ctx.Done():
		return ctx.Err()
	case err := <-errC:
		return err
	}
}
```

Reply:
Yes, that looks good! Added in [9f1a280](https://github.com/stackrox/scanner/pull/1118/commits/9f1a28099d6e1318fd66e3edf21c6894f6f6fe29)